File: python_version/main_conversion.py
#!/usr/bin/env python3

import sys
sys.path.append('/opt/src')

from python_version import set_nii_hdr as tools
from python_version import load_ismrmrd_ifft3d_reconstruction as h5reco
from python_version import extract_ismrmrd_parameters_from_headers as param
from python_version import flip_image as fi

import numpy as np
import nibabel as nib
import os
import subprocess
import re
import getopt


def main(argv):
    inputfile = ''
    outputfile = ''
    try:
        opts, args = getopt.getopt(argv,"hi:o:",["ifile=","ofile="])
    except getopt.GetoptError:
        print('test.py -i <inputfile> -o <outputfile>')
        sys.exit(2)
    for opt, arg in opts:
        if opt == '-h':
            print('ismrmrd_to_nifti-python -i <inputfile> -o <outputfile>')
            sys.exit()
        elif opt in ("-i", "--ifile"):
            inputfile = arg
        elif opt in ("-o", "--ofile"):
            outputfile = arg


    ## Load selected file
    filename = inputfile
    
    [pathstr, name] = os.path.split(filename)
    [name, ext]=os.path.splitext(name)

    if outputfile == '' :
        output_path = os.path.join(pathstr +'/'+ name +'_ismrmrd_to_nifti_version_python.nii')
    else :
        output_path = outputfile 
    print('Input file : ', filename)
    print('Output file : ', output_path)

    ## Reconstruct image (z,y,x)
    [head, hdr, img_scaled] = h5reco.load_ismrmrd_ifft3d_reconstruction(filename)

    ## Create parameters for set_nii_hdr et xform_mat
    h = param.extract_ismrmrd_parameters_from_headers(head, hdr)

    ## Get crop image, flip and rotationate to match with true Nifti image
    img = fi.flip_image(img_scaled[0])

    ## Create nii struct based on img
    nii_empty = nib.Nifti1Image(img, np.eye(4))
    pf = {"lefthand":0}
    h2 =[]
    h2.append(h)

    ## Compute nifti parameters
    [nii_filled, h3] = tools.set_nii_hdr(nii_empty, h2, pf)

    ## Save image in nifti format
    nib.save(nii_filled, output_path)

    print("Nifti well saved")


if __name__ == "__main__":
   main(sys.argv[1:])

File: raw_h5_to_nii/matlab_version/main_conversion.m
%% Add path to needed librairies

addpath('/usr/local/share/ismrmrd/matlab/')
addpath('/.../ismrmrd_to_nifti')
addpath('/.../ismrmrd_to_nifti/xiangruili-dicm2nii-b76a158')
clearvars


%% #################### SELECT A OUPUT PATH ###############################

% If output_path = '' : default path = data_path/filename+_ismrmrd_to_nifti_version_matlab.nii

output_path = ''; % need a .nii extension



 














%% #########################################################################
% ################## DON'T CHANGE ANYTHING FROM HERE ######################
% #########################################################################


[name, pathname] = uigetfile({'*.h5;*.mat','HDF5 File, MATLAB File (ISMRMRD format) (*.h5,*.mat)'},'MultiSelect','off');
filename = fullfile(pathname,name);


% ##################### LOAD FILES AND CONVERT DATA #######################

% Load selected file
[pathstr, name, ext] = fileparts(filename);
if strcmp(ext, '.h5')
    %load .h5 file and return dataset header, ismrmrd header and reconstructed
    %images
    [head, hdr, img_scaled] = load_ismrmrd_ifft3d_reconstruction(filename);
elseif strcmp(ext, '.mat')
    %load .mat file and return dataset header, ismrmrd header and reconstructed
    %images
    [head, hdr, img_scaled] = load_reconstructed_data(filename);
end


% Create parameters for set_nii_hdr et xform_mat
% needs to be done for each image (here size(img_scaled) = 1 so we work only
% on the first and only image of the dataset (in temporal dimension))
h = extract_ismrmrd_parameters_from_headers(head, hdr);

% Get crop image, flip and rotationate to match with true Nifti image
img = ismrmrd_to_nifti.flip_image(img_scaled{1});

% Create nii struct based on img
nii_empty = nii_tool('init', img); 

% In case there will be more that one temporal image in dataset :  h2 = h, with
% h a structure of size size(img_scaled)
% here, only one temporal image so h2{1} only
h2{1}=h;
pf.lefthand = 0; %to include in a function ?
[nii_filled, h3] = ismrmrd_to_nifti.set_nii_hdr(nii_empty, h2, pf);

fmt = 1;

% Choose to save nifti dataset in one unique file or in seperate (for 4D)
rst3D = (isnumeric(fmt) && fmt>3) || (ischar(fmt) && ~isempty(regexpi(fmt, '3D')));

% Save image in nifti format
if strcmp(output_path,'')
    nii_tool('save', nii_filled, [pathstr '/' name '_ismrmrd_to_nifti_version_matlab.nii'], rst3D);
else
    nii_tool('save', nii_filled, output_path, rst3D);
end
